src/sentry/models/environment.py

```diff
@@ -88,11 +88,17 @@ def get_or_create(cls, project, name):
         return env
 
     def add_project(self, project):
-        try:
-            with transaction.atomic():
-                EnvironmentProject.objects.create(project=project, environment=self)
-        except IntegrityError:
-            pass
+        cache_key = 'envproj:c:%s:%s' % (self.id, project.id)
+
+        envproj = cache.get(cache_key)
+        if env is None:
+            try:
+                with transaction.atomic():
+                    EnvironmentProject.objects.create(project=project, environment=self)
+            except IntegrityError:
+                pass
+            finally:
+                cache.set(cache_key, 1, 3600)
```

Comment:
I don't think we should put this in a `finally` in case another error prevented it from being created. Like, database being down or something.

I'd probably structure this so it sets it in the two conditions. Either it creates it successfully, or fails on IntegrityError. But we shouldn't cache otherwise.